File: MyGame/app/src/main/java/com/example/mygame/logic/CollisionHandler.kt
package com.example.mygame.logic

import com.example.mygame.`interface`.IGameObject
import com.example.mygame.`object`.Bullet
import com.example.mygame.`object`.Player
import com.example.mygame.`object`.Screen
import com.example.mygame.visitor.BulletCollisionVisitor
import com.example.mygame.visitor.PlayerCollisionVisitor
import com.example.mygame.visitor.ScreenCollisionVisitor

class CollisionHandler {

    fun checkCollisions(player: Player, screen: Screen, objects: List<IGameObject>) {
        checkScreenCollision(screen, objects)
        checkPlayerCollision(player, screen, objects)
        checkBulletCollision(objects)
    }

    private fun checkBulletCollision(objects: List<IGameObject>) {
        val bullet = objects.firstOrNull() { it::class == Bullet::class } as Bullet?
        if (bullet == null) {
            return
        }

        val bulletCollisionVisitor = BulletCollisionVisitor(bullet)

        objects.forEach {
            it.accept(bulletCollisionVisitor)
        }
    }

    private fun checkScreenCollision(screen: Screen, objects: List<IGameObject>) {
        val screenCollisionVisitor = ScreenCollisionVisitor(screen)

        objects.forEach {
            it.accept(screenCollisionVisitor)
        }
    }

    private fun checkPlayerCollision(player: Player, screen: Screen, objects: List<IGameObject>) {
        val playerCollisionVisitor = PlayerCollisionVisitor(player, screen.height)

        objects.forEach {
            if (it !is Player) {
                it.accept(playerCollisionVisitor)
            }
        }
    }
}

File: GameServer/src/Collision/CollisionHandler.php
<?php

namespace App\Collision;

use App\Model\Player;
use App\Model\Screen;
use App\Collision\Visitor\ScreenCollisionVisitor;
use App\Collision\Visitor\PlayerCollisionVisitor;

class CollisionHandler
{
    private array $gameState;

    private Screen $screen;

    private PlayerCollisionVisitor $playerCollisionVisitor;

    private ScreenCollisionVisitor $screenCollisionVisitor;

    public function __construct(Screen $screen)
    {
        $this->gameState = [
            'player' => [],
            'platform' => [],
            'bonus' => [],
            'endZone' => [],
        ];

        $this->playerCollisionVisitor = new PlayerCollisionVisitor($screen);
        $this->screenCollisionVisitor = new ScreenCollisionVisitor($screen);
    }

    public function setState(array $newState): void
    {
        $this->gameState = $newState;
    }

    public function updateScreen(Screen $screen)
    {
        $this->screen = $screen;
    }

    public function checkCollisions(array $gameState): void
    {
        $this->setState($gameState);

        $this->checkScreenCollision();

        foreach ($this->gameState['player'] as $player) {
            $this->checkPlayerCollision($player);
        }
    }

    private function checkScreenCollision(): void
    {
        if (isset($this->gameState['player'], $this->gameState['platform'], $this->gameState['bonus'])) {
            $objects = array_merge($this->gameState['player'], $this->gameState['platform'], $this->gameState['bonus']);
            foreach ($objects as $object) {
                $object->accept($this->screenCollisionVisitor);
            }
        }
    }

    private function checkPlayerCollision(Player $player): void
    {
        $this->playerCollisionVisitor->setPlayer($player);
        if (isset($this->gameState['player'], $this->gameState['platform'], $this->gameState['bonus'], $this->gameState['endZone'])) {
            $objects = array_merge($this->gameState['player'], $this->gameState['platform'], $this->gameState['bonus'], $this->gameState['endZone']);
            foreach ($objects as $object) {
                if (!$object instanceof Player) {
                    $object->accept($this->playerCollisionVisitor);
                }
            }
        }
    }
}
